source/stylist/fortran.py

```diff
@@ -392,3 +393,67 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
 
         issues.sort(key=lambda x: (x.filename, x.line, x.description))
         return issues
+
+
+class AutoCharArrayIntent(FortranRule):
+    """
+    Checks that all automatically assigned character arrays used as
+    subroutine or function arguments have intent(in) to avoid writing
+    outside the given array.
+    """
+    def __init__(self):
+        pass
+
+    def message(self, name, intent):
+        return (f"Arguments of type character(*) must have intent IN, but "
+                f"{name} has intent {intent}.")
+
+    def examine_fortran(self, subject: FortranSource) -> List[Issue]:
+        issues = []
+
+        # Collect all variable declarations
+        declarations: List[Fortran2003.Base] = list(
```

Comment:
Since you are looking for `Fortran2003.Type_Declaration_Stmt` shouldn't your list be of those rather than `Fortran2003.Base`?